Fixed a condition for grouped fields not visible properly in Questionnaire responses

hrit2773 opened the pull request:
## Proposed Changes

- Fixes #11207 

![Screenshot 2025-03-16 175339](https://github.com/user-attachments/assets/26c07331-99c4-4614-9bc1-4d427b09aa01)

@rithviknishad @Jacobjeevan 

## Merge Checklist

- [x] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [x] Ensure that UI text is kept in I18n files.
- [x] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [x] Request for Peer Reviews
- [ ] Completion of QA in Mobile Devices
- [ ] Completion of QA in Desktop Devices


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced the consultation questionnaire display so that grouped response fields now appear even when answers exist only in nested questions. This improvement ensures that all relevant details are clearly presented for a more comprehensive review of your consultation data.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hrit2773 committed:
fixed a condition for grouped fields not visible properly

hrit2773 commented:
@Jacobjeevan @rithviknishad if the responses contains responses for only type= group fields then the function was not propagating forward. So if the question further contains a questions field (group fields like Blood pressure containing questions like systolic and diastolic) then the function should propagate forward even if hasResponses is False we cant just stop there and return null. So hence I added an extra check for group.questions also. Can you review this. This solve the problem.

hrit2773 committed:
Merge branch 'develop' into bug

Jacobjeevan commented on src/components/Facility/ConsultationDetails/QuestionnaireResponsesList.tsx:
```diff
@@ -119,11 +117,10 @@ function QuestionGroup({
   );
 
-  if (!hasResponses) return null;
+  if (!hasResponses && !group.questions) return null;
```
Hmm, this isn't a good solution since you are just checking if group has sub questions, rather than checking for matching responses.

Instead you want to create a function that would do similar function as line 117, but also traverse through/check against sub questions when the question type is a group (keep in mind, you may have sub question within sub question...so forth), so you need to handle that.

hrit2773 replied:
@Jacobjeevan That condition is an exit condition of that recursive like loop. If we dont find any responses for the first time no worries the function can move forward and explore its questions and that happens so forth till we end up realizing that group.questoins is also not there and it doesnt have any responses yet so return null at that time. I guess this is a safe solution any other solution may interrupt the logic I guess...... what you have mentioned even this logic works like that under the hood

hrit2773 replied:
@Jacobjeevan thoughts??

hrit2773 commented:
@Jacobjeevan I have added a function that recursively checks the questions and finds id. This is also working. Can you just view it. I will discard the lock file changes

hrit2773 committed:
added extra function

hrit2773 commented:
@Jacobjeevan made the change

hrit2773 committed:
Merge branch 'develop' into bug

hrit2773 committed:
linting issue

hrit2773 commented:
@Jacobjeevan @rithviknishad a review on this?? I added a new function which would check for the nested questions also

hrit2773 commented:
@Jacobjeevan so like you asked me to add a function that would check for questions in questions and so forth.... I added a recursive function and I fixed the condition back to !hasResponses only as it was before. Just added some extra logic to check the sub questions having its responses or not. Its working fine. Can you just check the function.

hrit2773 committed:
Merge branch 'develop' into bug

Jacobjeevan commented on src/components/Facility/ConsultationDetails/QuestionnaireResponsesList.tsx:
```diff
@@ -127,15 +126,22 @@ function QuestionGroup({
   }[];
   level?: number;
 }) {
+  const checkResponse = (group: Question, r_id: string) => {
```
Should be a cleaner way to do this, something like:

<img width="559" alt="image" src="https://github.com/user-attachments/assets/eabf7cb2-5934-41c0-8656-45f2678d5485" />

hrit2773 replied:
Hmm... that would be ok as well does the same thing. Let me update it immediately

hrit2773 committed:
Merge branch 'develop' into bug

hrit2773 committed:
updated recursive function

hrit2773 commented:
@Jacobjeevan updated the function and verified. Its working !! I guess we are good to go. Further review cc @rithviknishad

hrit2773 committed:
Merge branch 'develop' into bug

hrit2773 committed:
added comments

Jacobjeevan commented on src/components/Facility/ConsultationDetails/QuestionnaireResponsesList.tsx:
```diff
@@ -127,15 +126,21 @@ function QuestionGroup({
+    // Recursively check if a question or any of its nested sub-questions have responses
+    // This ensures grouped fields are displayed even when only sub-questions have responses
+    if (question.type === "group") {
+      return question.questions?.some((q) => getHasResponse(q) ?? false);
```
false should be outside btw

hrit2773 replied:
Yup

hrit2773 commented:
@Jacobjeevan Changed it you can view it

hrit2773 committed:
discarded lock changes

hrit2773 commented:
@rithviknishad can you review this?

hrit2773 commented:
@nihal467 I guess we need another approving review and test to merge. @rithviknishad

nihal467 commented:
LGTM
